### PR TITLE
GH12: Grab clipboard image & upload file command

### PR DESCRIPTION
**Pull request by devlead**
Still work in progress but will
* upload to azure blob storage
* create a sas token (so container can be private, but image be accessible)
* remove temporary local image

Default container is `clipimg-vscode`, but it can be configured through setting `clipImg.storageAccountContainer`, this maybe should be asked for too.

Storage account will be stored in workspace settings and storage key will be stored securely using [keytar](https://github.com/atom/node-keytar) which on macOS the passwords are managed by the Keychain, on Linux they are managed by the Secret Service API/libsecret, and on Windows they are managed by Credential Vault.

First time command is executing in a workspace it'll ask for storage account name and key (if key not already in credential store).

### CODE REVIEW

**Comment by devlead** [comment]
@gep13 this should be ready for review now.

**Review by gep13 — COMMENTED** [review]
Just one question, so that I know what will happen...

**Comment by gep13** [comment]
@devlead would it be possible to add some docs to the readme to explain what needs to be done to get this all to work?

**Comment by devlead** [comment]
> @devlead would it be possible to add some docs to the readme to explain what needs to be done to get this all to work?

Yes it should.

File: src/blobstorage-service.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    await container.createIfNotExists();
+    let now = new Date();
+    let blobName = `${now.getFullYear()}/${(now.getMonth() + 1).toString().padStart(2, '0')}/${now.getDate().toString().padStart(2, '0')}/${imageFileName}`;
+    let blobClient = container.getBlockBlobClient(blobName);
+
+    await blobClient.uploadFile(imageFullFileName);
+    let startsOn = new Date();
+    let expiresOn = new Date();
+    startsOn.setDate(startsOn.getDate() - 1);
+    expiresOn.setFullYear(expiresOn.getFullYear() + 10);
```
[the remark this answers is not in the document]
Comment (devlead) [reply src/blobstorage-service.ts:31]: The link will stop working, it'll also stop working if they reset the storage key.

But files are all still there so you would need to create a new token and update markdown.

**Comment by devlead** [comment]
@gep13 README updated to reflect extension usage.

**Comment by devlead** [comment]
Probably should document settings to, let me know if you want in this or separate PR.

File: src/blobstorage-service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (gep13) [reply src/blobstorage-service.ts:31]: I take it we don’t _need_ the SAS token. With the correct settings in the storage account the files could be accessed anonymously, right? To that end, should we have an option to not include the SAS token to the URL?

[the remark this answers is not in the document]
Comment (devlead) [reply src/blobstorage-service.ts:31]: That could certainly be an option, as default I would prefer container not to be wide open.

[the remark this answers is not in the document]
Comment (devlead) [reply src/blobstorage-service.ts:31]: Another option along same would be to have a different base url if CDN attached to storage.

So perhaps an option to to have some kind of template for uri would be best?

**Review by gep13 — APPROVED** [review]
LGTM!

**Comment by gep13** [comment]
Thanks for getting this done! I think that there are a couple of follow up items that would be great to get done, but nothing to stop progressing with the PR.

**Pull request merged**